### PR TITLE
[6.0] Make LengthAwarePaginator respect distinct columns

### PR DESCRIPTION
**Pull request by erikgaal**
## Overview 
This PR demonstrates the bug that was reported in #21242. There have been _many_ previous PR attempts (#18186, #21899, #23209, #23214, #23255, #25126, #25973, #26014, #27107) to solve this issue, however, none succeeded.

## Reproduction
When using `paginate` on a `Builder` instance that has a `distinct` column, the actual amount of results from `$query->get()->count()` differ from `$query->paginate()->total()`.

_The issue has also been reproduced in `tests/Integration/Database/EloquentPaginateTest.php`._

## Proposed Solution
In #26014 it was proposed to allow to specify which columns the distinct should be on, and this PR implements that proposal. This solution gives a flexible way of building the query while not breaking existing behaviour.

The `distinct` property on `Builder` now can be a boolean or an array. When an `array` is passed, these denote the columns to be used in the distinct. If an argument is omitted or `true` is passed, the original behaviour should be used.

## Considerations
1. One other solution considered was to pass the columns from `$query->select()` to be directly passed to `getCountForPagination`, however, this would break existing behaviour and not work in SQLite when passing a wildcard or multiple columns to the select.

```php
$query->paginate(15, ['users.*'])
// Illuminate\Database\QueryException : SQLSTATE[HY000]: General error: 1 near "*": syntax error (SQL: select count(distinct "users".*) as aggregate from "users" inner join "posts" on "posts"."user_id" = "users"."id")
```

### CODE REVIEW

**Comment by stancl** [comment]
If you use the splat operator instead of using `func_get_args()`, an IDE won't shout at you if you pass multiple arguments.

```php
/**
 * Force the query to only return distinct results.
 *
 * @param array|mixed $columns,...
 * @return $this
 */
public function distinct(...$columns)
{
    $columns = count($columns) === 1 ? $columns[0] : $columns;
    $this->distinct = $columns;
    return $this;
}
```

This should work with all `('foo')`, `('foo', 'bar')` and `(['foo', 'bar'])`.

**Comment by taylorotwell** [comment]
Needs a very thorough explanation of why this is different from all other broken PRs on this issue.

**Comment by erikgaal** [comment]
@stancl 
> If you use the splat operator instead of using `func_get_args()`, an IDE won't shout at you if you pass multiple arguments.

I'd rather have the developer pass the columns as an array, but provide the `func_get_args()` as a legacy way to support the behaviour like in `select`.

@taylorotwell 
> Needs a very thorough explanation of why this is different from all other broken PRs on this issue.

I have provided multiple tests that show it 1. doesn't break any current normal behaviour (without distinct), 2. that it still supports the currently _broken_ behaviour (difference between `get()->count()` and `paginate()->total()`) and 3. the behaviour we would like to see (uniform counts). What other proof would you require that it doesn't break?

**Comment by deleugpn** [comment]
This is pretty good overall. It doesn't break the test added on https://github.com/laravel/framework/pull/23259 and it nicely implements the idea from @stayallive [https://github.com/laravel/framework/pull/26014]. We will probably need a good documentation to explain why `->get()->count()` will differ from `paginate()->total()`, but since this bug is a few years old, I think it was a nice alternative to not change the behavior if anybody is relying on the broken behavior for some unknown reason. "An old enough bug is a feature" 🤣 

The only thing I'm unsure about is how `SELECT DISTINCT(one_column), another_column` plays on a MySQL environment with `strict` enabled. I know for Group By MySQL will scream at you saying that you are retrieving something outside of the aggregation clause. 

Aside from that I think this is pretty neat.

@taylorotwell the biggest difference here from all other proposed solutions is the fact that `distinct` is no longer a boolean. It now carries over the columns which you want to specify as `DISTINCT`, which allows the paginator to count distinct values based only on the specified columns.

**Comment by sisve** [comment]
There are some changes to property types and method signatures that would qualify as breaking changes. Are you sure this should target the current release and not the next one?

**Comment by rovansteen** [comment]
Since the methods aren't typed and this change is backward compatible I don't see why you would qualify this as a breaking change.

**Comment by staudenmeir** [comment]
It's a breaking change when you override the `distinct()` method in your own query builder.

**Comment by deleugpn** [comment]
The PR was proposed before 5.8 release and 5.8 was released yesterday. I think it's fine to land this on 5.8.

**Comment by rovansteen** [comment]
@staudenmeir since it's not typed it won't break. The current implementation expects a boolean and this change still supports that. It would be more constructive if you could give an example of a piece of code that works right now but would break with this change.

**Comment by staudenmeir** [comment]
```php
namespace App;

class Builder extends \Illuminate\Database\Query\Builder {
    public function distinct() {
        //
    }
}

(new \App\Builder(\DB::connection()))->distinct();
```

> Declaration of App\Builder::distinct() should be compatible with Illuminate\Database\Query\Builder::distinct($column = true)

**Comment by sisve** [comment]
@deleugpn If it is merged, it merges into a released branch. The important part here isn't the state when the PR started, but the state when the PR is merged.

**Comment by driesvints** [comment]
This definitely counts as a breaking change atm.

**Comment by rovansteen** [comment]
@driesvints it feels kind of silly to wait 6 months for 5.9 to release this **bug fix**. But if you insist on not shipping this until the next version we could change the distinct method to:
```php
public function distinct()
{
    $args = func_get_args();

    if (count($args) === 0) {
        // Current behavior
        $this->distinct = true;
    } else {
        // New behavior
        $this->distinct = is_array($args[0]) || is_bool($args[0]) ? $args[0] : $args;
    }

    return $this;
}
```
And then we can change the function signature for 5.9.

**Comment by stancl** [comment]
@rovansteen 

It feels very wrong to have most of the method's behavior depend on the arguments while not expecting a single one.

What benefit is there to using `func_get_args` in PHP 7, when you can use the splat operator?

Your snippet will work exactly the same if you remove the `$args = func_get_args()` line and add an `...$args` argument.

I agree that it feels silly to wait 6 months to ship this.

Since this is a bug fix and 5.8 is only like two days old, the upgrade guide could have a section for upgrading from 5.8.0 to 5.8.x.

The 5.6 upgrade guide does this https://laravel.com/docs/5.6/upgrade.

However that is a security release and this is just a bug fix. Depends on how important this bug fix is deemed.

**Comment by erikgaal** [comment]
@stancl 
I think the entire point of @rovansteen was to prevent changing the signature of the `distinct` method. Adding the `...$args` will still cause a conflict when someone is extending the `Builder` class.

**Comment by rovansteen** [comment]
@stancl this PR is about fixing an issue with distinct in combination with the paginator, let's stay on topic. I would suggest opening a PR yourself to change `func_get_args` to the splat operator in the codebase so we can separate the discussions.

**Comment by stancl** [comment]
That was only one point of mine, the second was about when to release this. And thanks for the clarification @erikgaal about not changing the signature of the `distinct` method, I misunderstood @rovansteen's comment. I thought he meant shipping his snippet in 5.9.

**Comment by deleugpn** [comment]
We've had this problem for a long time now. If this solution is accepted by it's merits, I don't think we should wait 6 months to get this fix just because of 2 days of Laravel 5.8 release. I doubt somebody rushed into overriding this method between the release of 5.8 and now.
If the problem is the breaking change, we can open 2 PRs: one with @rovansteen's solution for 5.8 and re-target this PR to 5.9.

**Comment by stancl** [comment]
> I doubt somebody rushed into overriding this method between the release of 5.8 and now.

Was the method changed between 5.7 and 5.8? While it's unlikely that someone rushed into extending this method in the last 2 days, it's possible that someone extended it in previous versions and updated to 5.8 in the last 2 days.

I think there's no reason to make an unexpected (however minor) breaking change if there is a backwards-compatible solution (@rovansteen's).

**Comment by devcircus** [comment]
I know everyone seems convinced this is finally the pr that's going to work, however this problem has always been in the framework and every pr to fix it has eventually been reverted. As devs have worked around this for years, any fix could cause problems with existing code. To me, it makes more sense to merge into 5.9 and give it the opportunity to be used and tested thoroughly before we go live with it. As the framework has always had the current behavior, there is no reason to rush it through and break apps.

**Comment by taylorotwell** [comment]
Hmm, I have to admit on reviewing the code I don't see a way this can break existing applications other than the signature change. If we want to do it on 5.8 I think the **safest** route for us is probably to not change the signature and use func_get_args() to inspect the arguments that are passed in. On 5.9 we can clean that up explicitly add the argument.

**Comment by jimgwhit** [comment]
I have never had any trouble, but when using LengthAwarePaginator I use getPdo(). Which is probably better anyway if using LengthAwarePaginator.

**Comment by sisve** [comment]
I also believe that the change of type of the $distinct property is a breaking change. This can probably be circumvented by having a new property for the array with column names.

Using `func_get_args` here to avoid an signature change would be like telling people to turn off warnings. It will hide the problem, but still cause breaking changes. A composer package that fully supports 5.8 will need similar hacks to handle the _documented_ change in behavior of the method. The method signature alone is only a part of the problem. As an extreme example; if we document that the distinct() method will play a small song on the pc speaker, then it would be a breaking change because everyone that is overriding the method needs to change their published code to match these new expectations, even if the signature itself didn't change.

The code implies that I should be able to call `->distinct('title', 'user_id')` to have the result distinct by two fields, but there's no tests that checks this. When I try I get a `SQLSTATE[HY000]: General error: 1 row value misused (SQL: select count(distinct ("title", "user_id")) as aggregate from "posts")`. I think this is an easy fix, the paranthes for distinct shouldn't be  there. It should be `select count(distinct "title", "user_id") ...`

**Comment by jimgwhit** [comment]
@sisve Are you sure there is a problem with distinct?  If people write a query, and don't use 
distinct correctly, of course it seems like there's something wrong.  Anyway it works correctly using getPdo().

**Comment by devcircus** [comment]
As mentioned, the bug has existed for years. There's no reason to shoehorn it into the framework with hacks to try to avoid breaking changes. Also, because we've worked around this bug for years, some apps are bound to have unexpected breaks either way. To me, it makes sense to merge a clean version, without any tweaks to avoid breaking changes, into 5.9, giving us 6 months to uncover any issues.

**Comment by jimgwhit** [comment]
Here's a larger query for a report, you only need to count once, if you use session to store count.
```
    public function report()
    {

        $page = Request::input('page', '1');
        $bdate = Request::input('begindate');
        $edate = Request::input('enddate');
               
        $perpage = "10";
        $offset = ($page - 1) * $perpage;

        if (!$page === '1') {
            Session::get('kount');
        } else {
            $krows = "SELECT COUNT(*) AS kount FROM ";
            $krows .= "(SELECT distinct `account_types`.`AccountType` AS `AccountType`,`accounts`.`AccountNumber` AS `AccountNumber`,`accounts`.`AccountName` AS `AccountName`,sum(`transactions`.`Expense`) AS `Sum_Expense`,sum(`transactions`.`Income`) AS `Sum_Income` from ((`account_types` join `accounts` on((`account_types`.`AccountTypeID` = `accounts`.`AccountTypeID`))) join `transactions` on((`accounts`.`AccountID` = `transactions`.`AccountID`))) where (`transactions`.`TransactionDate` Between '$bdate' and  '$edate') group by `account_types`.`AccountType`,`accounts`.`AccountNumber`,`accounts`.`AccountName`) as t1";
            $sthk = DB::getPdo()->prepare($krows);
            $sthk->execute();
            $quyk = $sthk->fetch(\PDO::FETCH_OBJ);
            $numrows = $quyk->kount;
        }

        if ($page = '1') {
            Session::put('numrows', $numrows);
        }

        $numrows = Session::get('numrows');

        $pagingQuery = "LIMIT {$offset}, {$perpage}";

        $sql = "select distinct `account_types`.`AccountType` AS `AccountType`,`accounts`.`AccountNumber` ";
        $sql .= "AS `AccountNumber`,`accounts`.`AccountName` AS `AccountName`,sum(`transactions`.`Expense`) ";
        $sql .= "AS `Sum_Expense`,sum(`transactions`.`Income`) AS `Sum_Income` ";
        $sql .= "from ((`account_types` join `accounts` on((`account_types`.`AccountTypeID` = ";
        $sql .= "`accounts`.`AccountTypeID`))) join `transactions` on((`accounts`.`AccountID` = `transactions`.`AccountID`))) ";
        $sql .= "where (`transactions`.`TransactionDate` Between '$bdate' and  '$edate') ";
        $sql .= "group by `account_types`.`AccountType`,`accounts`.`AccountNumber`,`accounts`.`AccountName` ";
        $sql .= $pagingQuery;

        $sth = DB::getPdo()->prepare($sql);
        $sth->execute();
        $quy = $sth->fetchAll(\PDO::FETCH_OBJ);
        $report = new LengthAwarePaginator($quy, $numrows, $perpage);
        $pagelinks = ['bdate' => $bdate, 'edate' => $edate, 'page' => $page];
        
        $title = 'Monthly Report';
        return view('account.report', compact('report', 'title'));
    }

```
If folks would just use PDO getPdo() with LengthAwarePaginator.

And view:

```
@php echo str_replace('/?', '?', $report->appends($pagelinks)->render())  @endphp
```

Just that easy.

**Comment by taylorotwell** [comment]
@sisve I'm not sure I quite follow. If I'm a package that is choosing to override the `distinct` method, I've already chosen to ignore whatever the framework is doing, including any improvements or changes it might make, right?

**Comment by sisve** [comment]
@taylorotwell It's true that the implementer steps away from the framework's code (ignoring calls to `parent::distinct(...)`), but they are still expected to follow the documented behavior and signature of the method. These changes imply a documentation change stating that the method can be passed additional arguments to do something more than the usual stuff. People will start using this, and then file bug reports to package developers that state their packages fully support 5.8 doesn't work as expected. This is because the framework has changed the expectations of the method, during the lifetime of the release.

I understand that my view on this is rather strict. This is a breaking change, but I want to be clear that solely being a breaking change will not stop it from being merged into an active release. Many bugfixes will contain breaking changes in some form; and in some cases people are depending on the bugged result. If this is merged, make sure we document it in the changelog to let package developers know about changes. Slightly relevant xkcd: https://xkcd.com/1172/

Regarding the exception when doing `->distinct('title', 'user_id')`, it's the new code in this PR that adds the erroneous parentheses in the compileColumns method. The parentheses after the distinct keyword isn't an invocation, but affects the argument, similar to `$var = ('string')`. Having two values within them does not work. The correct syntax is `COUNT(DISTINCT col1, col2, ...)`

**Comment by jimgwhit** [comment]
You folks aren't going to mess up the length aware paginator so it won't work with regular queries are you.

**Comment by stancl** [comment]
toArray() works with both and there is a separate thread (#27722) to discuss that.

**Comment by taylorotwell** [comment]
So there is still a problem with `distinct` in this PR?

**Comment by erikgaal** [comment]
@sisve 

You're describing two different things there; `distinct` without parameters means something else than with parameters. See the tests below.

https://github.com/laravel/framework/blob/1ea14eeb198489b3037466bcfa16702b11f57cd8/tests/Database/DatabaseQueryBuilderTest.php#L115-L127

If you want to `distinct` on all columns, you should call distinct without parameters. Not all DMS's support `COUNT(DISTINCT (col1, col2))`, that's true, but some like Postgres do. If you're using distinct with multiple columns when your DMS doesn't support it then that's your own fault.

**Comment by sisve** [comment]
@erikgaal

> Not all DMS's support COUNT(DISTINCT (col1, col2)), that's true, but some like Postgres do.

Ah, so here's a very important thing; the syntax differs between engines. Postgresql requires the parameters, mysql does not support them. I was only testing on mysql and made the assumption the syntax for multiple columns wasn't tested.

    // postgresql:
    SELECT COUNT(DISTINCT (col1, col2)) FROM tbl

    // mysql:
    SELECT COUNT(DISTINCT col1, col2) FROM tbl

    // sqlite, sql server:
    (not supported)

So, to me, this looks like it's generating bad sql for mysql. The query grammar for mysql needs to override compileAggregate to remove those parentheses.

**Comment by taylorotwell** [comment]
@erikgaal: any thoughts on @sisve's latest comments?

**Comment by erikgaal** [comment]
@sisve

I've digged into the syntaxes for distinct queries a bit more and came to the following conclusions.

1. Only Postgres supports distincting _on_ a subset of columns from those that are selected. Which raises the the question if we should still allow the `Builder` to be passed a set of columns, even though this behaviour is only supported in Postgres? Must it be ignored in other grammars? 

```php
$query->distinct('title')->from('posts')->get();
```

```sql
-- Postgres
select distinct on ("title") * from "posts";

-- MySQL
select distinct * from "posts";
``` 

2. In the case for aggregates, MySQL and Postgres (and other grammars) share the syntax as far as I know.

```php
$query->distinct('title')->from('posts')->count();
```

```sql
-- Postgres and MySQL
SELECT COUNT(DISTINCT "title") FROM "posts";
```

I've made some changes to make sure the queries are correct for the used grammars.

**Comment by deleugpn** [comment]
The Builder is essentially a wrapper around SQL and it's very likely that there's combination of features that would result in an invalid SQL statement. As long as it doesn't silently fail, I think we're good.
I think for MySQL users the wish is exactly to use distinct with count, an aggregator, which is the whole point of providing this to fix the Paginator

**Comment by erikgaal** [comment]
Hi @taylorotwell and team,

What's the direction you would like to see this PR go to?

**Comment by narwy** [comment]
> So there is still a problem with `distinct` in this PR?

I agree with @sisve and see this as a breaking change. Should be rolled out in 5.9.

**Comment by devcircus** [comment]
Since we're about 3 months to 5.9, is have to assume that is the plan.

**Comment by driesvints** [comment]
@devcircus probably 5 months because it's likely we'll launch in september. Also see: https://blog.laravel.com/laravel-release-schedule

File: src/Illuminate/Database/Query/Grammars/Grammar.php [hunk 1]
```diff
@@ -109,7 +109,9 @@ protected function compileAggregate(Builder $query, $aggregate)
         // If the query has a "distinct" constraint and we're not asking for all columns
         // we need to prepend "distinct" onto the column name so that the query takes
         // it into account when it performs the aggregating operations on the data.
-        if ($query->distinct && $column !== '*') {
+        if (is_array($query->distinct)) {
+            $column = 'distinct '.$this->columnize($query->distinct).'';
```
Comment (danilopinotti) [inline src/Illuminate/Database/Query/Grammars/Grammar.php]: Why concatenate with an empty string at the end??

**Comment by taylorotwell** [comment]
Can this be re-targeted to 5.9?

**Comment by erikgaal** [comment]
@taylorotwell Fine by me

**Comment by taylorotwell** [comment]
Well here we go 😆 ❤️ 

@erikgaal hopefully you are the lucky winner to solve this! :)

**Pull request merged**